Merge develop branch

NowellPhelps opened the pull request:
Implemented changes from HW2

Wenxuan-Lu commented on tests/testthat/helper.R:
-1pt; All exported functions should under /R. Only functions that are necessary for user need to be exported. Functions under /tests should not be visible to the user.

NowellPhelps replied:
Corrected

Wenxuan-Lu commented on tests/testthat/test-are_all_close.R:
In testthat, it is recommended to have one test case per function. Instead of separating test cases with comments, please organize them into separate test_that() function calls.

NowellPhelps replied:
Corrected

Wenxuan-Lu commented:
-2pts; Exported functions should be documented using Roxygen2-style comments to automatically generate man files.

Wenxuan-Lu commented on R/hiperglm.R:
It is recommended to use the function header hiper_glm <- function(design, outcome, model = "linear", option = list()) as shown in Aki's repo to ensure easier extension for future implementations.

NowellPhelps replied:
Implemented

Wenxuan-Lu commented on R/hiperglm.R:
-1pt; Use one blank line to separate logical sections of code, but do not overuse it.

NowellPhelps replied:
I have reduced the amount of blank lines between code

Wenxuan-Lu commented on R/hiperglm.R:
Please check the validity of the input parameters. In this case, terminate the program if the parameter optimizer is invalid.

NowellPhelps replied:
Implemented

Wenxuan-Lu commented on R/linear_model_functions.R:
-1pt; Please leave a space on both sides of arithmetic operations for consistency.

Wenxuan-Lu replied:
-1pt; The function neg_log_likelihood does not seem to be used anywhere. Please remove functions that are never called.

Wenxuan-Lu commented on R/linear_model_functions.R:
```diff
@@ -0,0 +1,39 @@
+    warning("Incompatible sizes of design matrix and coefficient vector")
+  }
+
+  l <- -nrow(design)*log(2*pi*sigma_squared)/2 - (1/(2*sigma_squared)) * t(outcome - design%*%beta)%*%(outcome - design%*%beta)
```
The constant part can be removed since it does not affect the result.

Wenxuan-Lu commented on R/linear_model_functions.R:
```diff
@@ -0,0 +1,39 @@
+# '@export
+gradient_linearmodel <- function(beta, design, outcome, sigma_squared = 1){
+
+  if(!(nrow(design) == length(outcome))){
```
Nice error checking on the input parameters. Moving it to the main function hiperglm() may be more beneficial.

Wenxuan-Lu commented on R/hiperglm.R:
Great job writing the log-likelihood and gradient log-likelihood as separate functions in another file. Though not necessary for this implementation, moving the actual calculation steps (i.e., lines 5 and 8-13) to separate functions in a separate file would be helpful. This will make the code more modular and flexible, especially for developing a generic function that can accommodate various models and fitting methods.

Wenxuan-Lu commented on R/hiperglm.R:
```diff
@@ -1,5 +1,21 @@
+                        gr = function(par) -gradient_linearmodel(par, design, outcome, sigma_squared = 1),
+                        method = "BFGS")
+
+    coef <- mle$par
```
-1 pt; Please check whether BFGS has converged before returning the result. If it has not converged, provide a warning message to alert the user.

Wenxuan-Lu reviewed (changes requested):
Great job! I enjoyed reviewing your code. Remember that you can earn back all points by addressing these comments. Please address both point-deducting and non-point-deducting comments. I will update this comment with the final score when HW3 is due.

NowellPhelps commented:
> Great job! I enjoyed reviewing your code. Remember that you can earn back all points by addressing these comments. Please address both point-deducting and non-point-deducting comments. I will update this comment with the final score when HW3 is due.

Thank you Wenxuan! I have added my commits to the develop branch here. I think these changes should be visible to you but please let me know if not. Thank you for all the feedback on how to improve this code.
